File: exercises/domain-1/block-1.3/starter.py
from typing import Callable, Any
import asyncio

HookFn = Callable[[str, dict], bool | None]
StepFn = Callable[[dict], dict]


class WorkflowEngine:
    def __init__(self):
        self.steps: list[tuple[str, StepFn]] = []
        self.pre_hooks: list[HookFn] = []
        self.post_hooks: list[HookFn] = []

    def add_step(self, name: str, fn: StepFn) -> "WorkflowEngine":
        self.steps.append((name, fn))
        return self

    def add_pre_hook(self, hook: HookFn) -> "WorkflowEngine":
        # TODO: Add hook to self.pre_hooks
        return self

    def add_post_hook(self, hook: HookFn) -> "WorkflowEngine":
        # TODO: Add hook to self.post_hooks
        return self

    async def run(self, initial_context: dict = {}) -> dict:
        context = dict(initial_context)

        for name, step_fn in self.steps:
            # TODO: Run pre_hooks — if any return False, skip this step
            # TODO: Run the step (if not skipped)
            # TODO: Run post_hooks
            print(f"Running step: {name}")

        return context


# Demo: Document processing pipeline
engine = WorkflowEngine()

engine.add_step("read", lambda ctx: {**ctx, "document": "Hello World content", "word_count": 2})
engine.add_step("transform", lambda ctx: {**ctx, "document": ctx["document"].upper()})
engine.add_step("save", lambda ctx: {**ctx, "saved": True})

# TODO: Add a pre_hook that blocks "save" if word_count < 5
# TODO: Add a post_hook that logs step completion


async def main():
    result = await engine.run({"filename": "test.txt"})
    print("Final context:", result)


asyncio.run(main())

File: exercises/domain-1/block-1.3/starter.ts
type HookFn = (stepName: string, context: Record<string, unknown>) => boolean | void | Promise<boolean | void>

interface Step {
  name: string
  fn: (context: Record<string, unknown>) => Promise<Record<string, unknown>>
}

class WorkflowEngine {
  private steps: Step[] = []
  private preHooks: HookFn[] = []
  private postHooks: HookFn[] = []

  addStep(name: string, fn: Step["fn"]): this {
    this.steps.push({ name, fn })
    return this
  }

  // TODO: Implement addPreHook
  addPreHook(hook: HookFn): this {
    // Add to this.preHooks
    return this
  }

  // TODO: Implement addPostHook
  addPostHook(hook: HookFn): this {
    // Add to this.postHooks
    return this
  }

  async run(initialContext: Record<string, unknown> = {}): Promise<Record<string, unknown>> {
    let context = { ...initialContext }

    for (const step of this.steps) {
      // TODO: Run pre-hooks — if any return false, skip this step
      // TODO: Run the step (if not skipped)
      // TODO: Run post-hooks
      console.log(`Running step: ${step.name}`)
    }

    return context
  }
}

// Demo: Document processing pipeline
const engine = new WorkflowEngine()

engine
  .addStep("read", async (ctx) => {
    console.log("  → Reading document")
    return { ...ctx, document: "Hello World content", wordCount: 2 }
  })
  .addStep("transform", async (ctx) => {
    console.log("  → Transforming document")
    return { ...ctx, document: (ctx.document as string).toUpperCase() }
  })
  .addStep("save", async (ctx) => {
    console.log("  → Saving document")
    return { ...ctx, saved: true }
  })

// TODO: Add a pre-hook that blocks the "save" step if wordCount < 5
// TODO: Add a post-hook that logs each step completion

engine.run({ filename: "test.txt" }).then((result) => {
  console.log("Final context:", result)
})
